fetch-in-react-practice/client/components/app.jsx

```diff
@@ -0,0 +1,123 @@
+    *   - replace the old todos array in state with the new one
+    *
+    * DO NOT MUTATE the original state array, nor any objects within it.
+    * https://reactjs.org/docs/optimizing-performance.html#the-power-of-not-mutating-data
+    *
+    * TIP: Be sure to SERIALIZE the todo object in the body with JSON.stringify()
+    * and specify the "Content-Type" header as "application/json"
+    *
+    * TIP: Use Array.prototype.concat to create a new array containing the contents
+    * of the old array, plus the object returned by the server.
+    */
+    fetch('/api/todos', {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json'
+      },
+      body: JSON.stringify(newTodo)
+    })
+      .then(response => response.json())
+      .then(newTodo => {
+        const copy = this.state.todos;
+        const newArr = copy.concat(newTodo);
+        this.setState({ todos: newArr });
+      })
+      .catch(err => console.error('Fetch failed!', err));
+  }
+
+  toggleCompleted(todoId) {
+    /**
+     * Find the index of the todo with the matching todoId in the state array.
+     * Get its "isCompleted" status.
+     * Make a new object containing ONE PROPERTY: the opposite "isCompleted" status.
+     * Use fetch to send a PATCH request to `/api/todos/${todoId}`
+     * Then 😉, once the response JSON is received and parsed,
+     *   - create a shallow copy of the todos array from state
+     *   - replace the old todo with the todo received from the server
+     *   - replace the old todos in the state with the new one (you know the index).
+     *
+     * NOTE: "toggle" means to flip back and forth, so clicking a todo
+     * in the list repeatedly should "toggle" its isCompleted status back and forth.
+     *
+     * DO NOT try to calculate the index of the todo by subtracting 1 from the id.
+     *
+     * DO NOT MUTATE the original state array, nor any objects within it.
+     * https://reactjs.org/docs/optimizing-performance.html#the-power-of-not-mutating-data
+     *
+     * TIP: Be sure to SERIALIZE the updates in the body with JSON.stringify()
+     * And specify the "Content-Type" header as "application/json"
+     */
+    let current = null;
+    this.state.todos.forEach((object, index) => {
+      if (object.todoId === todoId) {
+        current = index;
+      }
+    });
+    const completed = this.state.todos[current].isCompleted;
+    const newObj = { isCompleted: !completed };
+
+    fetch(`/api/todos/${todoId}`, {
+      method: 'PATCH',
+      headers: {
+        'Content-Type': 'application/json'
+      },
+      body: JSON.stringify(newObj)
+    })
+      .then(response => response.json())
+      .then(data => {
+        const copy = this.state.todos;
+        const currentTodo = this.state.todos[current];
+        currentTodo.isCompleted = newObj.isCompleted;
+        copy.splice(current, 1, currentTodo);
+        this.setState({ todos: copy });
```

Comment:
Here since you know the `index` of the todo that needs updated from above you can simplify this a bit:
```JAVASCRIPT
copy[current] = data;

this.setState({ todos: copy });
```

Reply:
Thank you for all of the feedback!